### PR TITLE
Allow the user to configure the displayed precision

### PR DESCRIPTION
**Pull request by chas-iot**
- Temperature
 - Humidity
 - Pressure

Issues & notes
1 - I can't work out how the manifest is loaded and it is missing the id
     The manifest.id is needed to access the configuration, so I
     provide a default in a hacky way. Please could you take a look?
     Flagged with @tim
2 - After some research, I chose opinionated default precisions that
     should be relevant for most users and line up with what is
     reported by many weather services
3 - I convert from the two line 'setCachedValue' plus
     'notifyPropertyChanged' to the one line 'setCachedValueAndNotify'
     This means that no values are sent when the displayed value does
     not change. Which has huge benefits when using the logs and graphs
4 - After trying several alternatives, I perform the precision
     conversion using toFixed() to get a string value and prefixing with
     a + to convert back to numeric. Using multiplication, Math.round()
     and division often led to numbers with 12 or more 0s followed
     by a another digit
5 - Layering. I separated out the scaling to a new source file.
     It may needed to be added in package.sh etc

### CODE REVIEW

File: src/ruuvitag-adapter.ts [hunk 1]
```diff
@@ -122,17 +125,13 @@ export class RuuviTag extends Device {
       batteryVoltage
     } = data;
 
-    this.humidityProperty.setCachedValue(humidity);
-    this.notifyPropertyChanged(this.humidityProperty);
+    this.humidityProperty.setCachedValueAndNotify(scaleHumidity(humidity,this.config));
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts]: The method `setCachedValueAndNotify` was actually introduced by me 😬.
Using it breaks compatibility with gateways before version `0.10.0`.
There was no way to express this dependency until the introduction of the `manifest.json`.
That's why I didn't use it right away.
If we use it, at least the property `moziot.api.min` in the `package.json` needs to be removed.
I would propose to remove the whole `moziot` section in the `package.json` though.
It is only there to maintain compatibility with older gateways.

File: src/ruuvitag-adapter.ts [hunk 2]
```diff
@@ -157,6 +155,20 @@ export class RuuviTagAdapter extends Adapter {
     this.knownDevices = {};
     addonManager.addAdapter(this);
 
+    // @tim - the manifest that is loaded (I can't work out how) does not have the id
+    //        The id is required to load the configuration, so I hack it in here
+    //        Sometimes I've seen a generated manifest.json in the lib directory
+    manifest.id = manifest.id || 'ruuvitag-adapter';
+
+    const config = getDefaultConfig();
+    const db = new Database(manifest.id);
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts]: The `manifest` predates the `manifest.json` file.
That's why the structure is based on the `package.json` file.
Unfortunately, there is no way to ever change it without breaking existing code.
Instead of `id` and `name` its `name` and `display_name`.
```suggestion
    const db = new Database(manifest.name);
```

File: manifest.json [hunk 3]
```diff
@@ -15,5 +15,42 @@
...
+          "minimum":0,
+          "maximum":4
+        },
+        "pressurePrecision": {
+          "title": "Pressure precision",
+          "description": "The number of decimal places to show in atmospheric pressure",
+          "type": "integer",
+          "enum": [ 0, 1, 2 ],
+          "minimum":0,
+          "maximum":2
```
Comment (tim-hellhake) [inline manifest.json:45]:
```suggestion
          "enum": [ 0, 1, 2 ]
```

File: manifest.json [hunk 4]
```diff
@@ -15,5 +15,42 @@
...
+          "minimum":0,
+          "maximum":3
+        },
+        "humidityPrecision": {
+          "title": "Humidity precision",
+          "description": "The number of decimal places to show in humidity",
+          "type": "integer",
+          "enum": [ 0, 1, 2, 3, 4 ],
+          "minimum":0,
+          "maximum":4
```
Comment (tim-hellhake) [inline manifest.json:37]:
```suggestion
          "enum": [ 0, 1, 2, 3, 4 ]
```

File: manifest.json [hunk 5]
```diff
@@ -15,5 +15,42 @@
...
+    "schema": {
+      "type": "object",
+      "properties": {
+        "temperaturePrecision": {
+          "title": "Temperature precision",
+          "description": "The number of decimal places to show in temperature",
+          "type": "integer",
+          "enum": [ 0, 1, 2, 3 ],
+          "minimum":0,
+          "maximum":3
```
Comment (tim-hellhake) [inline manifest.json:29]: I don't think those are necessary if the enum property is set.

```suggestion
          "enum": [ 0, 1, 2, 3 ]
```

File: src/ruuvitag-scaling.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+export function getDefaultConfig() {
+    return {
+        temperaturePrecision: 1,
+        humidityPrecision: 0,
+        pressurePrecision: 0,
+        temperatureStep: 0.1,
+        humidityStep: 1,
+        pressureStep: 1,
+    };
+}
```
Comment (tim-hellhake) [inline src/ruuvitag-scaling.ts]: You need to add this file to the files array in the package.json.

File: src/ruuvitag-adapter.ts [hunk 7]
```diff
@@ -6,27 +6,30 @@
...
     super(adapter, `${RuuviTag.name}-${id}`);
     this['@context'] = 'https://iot.mozilla.org/schemas/';
     this['@type'] = ['TemperatureSensor'];
     this.name = `RuuviTag (${address || id})`;
     this.description = manifest.description;
+    this.config = config;
 
     const data = parse(manufacturerData);
-    const metadata = getMetadata(data.version);
+    const metadata = getMetadata(data.version,config);
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts]:
```suggestion
    const metadata = getMetadata(data.version, config);
```

File: src/ruuvitag-adapter.ts [hunk 8]
```diff
@@ -6,27 +6,30 @@
 
 'use strict';
 
-import { Adapter, Device, Property } from 'gateway-addon';
+import { Adapter, Device, Property, Database } from 'gateway-addon';
 
 import noble from '@abandonware/noble';
-import { parse, DataV3, DataV5, getMetadata } from './ruuvitag-parser';
+import { parse, DataV3, DataV5 } from './ruuvitag-parser';
+import { getMetadata, scaleTemperature, scaleHumidity,scalePressure, getDefaultConfig, mergeLoadedConfig } from './ruuvitag-scaling';
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts]:
```suggestion
import { getMetadata, scaleTemperature, scaleHumidity, scalePressure, getDefaultConfig, mergeLoadedConfig } from './ruuvitag-scaling';
```

File: src/ruuvitag-adapter.ts [hunk 9]
```diff
@@ -157,6 +155,20 @@ export class RuuviTagAdapter extends Adapter {
     this.knownDevices = {};
     addonManager.addAdapter(this);
 
+    // @tim - the manifest that is loaded (I can't work out how) does not have the id
+    //        The id is required to load the configuration, so I hack it in here
+    //        Sometimes I've seen a generated manifest.json in the lib directory
+    manifest.id = manifest.id || 'ruuvitag-adapter';
+
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts]: After changing to manifest.name you no longer need those.

```suggestion
```

File: src/ruuvitag-scaling.ts [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                    min: 1.6,
+                    max: 3.647,
+                    step: 0.001
+                }
+            };
+    }
+
+    throw `Unknown version ${version}`;
+}
+export function mergeLoadedConfig(current: any, loaded: any) {
```
Comment (tim-hellhake) [inline src/ruuvitag-scaling.ts]:
```suggestion
}

export function mergeLoadedConfig(current: any, loaded: any) {
```

File: src/ruuvitag-scaling.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+export function getDefaultConfig() {
+    return {
+        temperaturePrecision: 1,
+        humidityPrecision: 0,
+        pressurePrecision: 0,
+        temperatureStep: 0.1,
+        humidityStep: 1,
+        pressureStep: 1,
```
Comment (tim-hellhake) [inline src/ruuvitag-scaling.ts:97]: If I saw it correctly those values are never read because they are overwritten by `mergeLoadedConfig`.
```suggestion
        pressurePrecision: 0
```

File: src/ruuvitag-adapter.ts [hunk 12]
```diff
@@ -122,17 +125,13 @@ export class RuuviTag extends Device {
       batteryVoltage
     } = data;
 
-    this.humidityProperty.setCachedValue(humidity);
-    this.notifyPropertyChanged(this.humidityProperty);
+    this.humidityProperty.setCachedValueAndNotify(scaleHumidity(humidity,this.config));
 
-    this.temperatureProperty.setCachedValue(temperature);
-    this.notifyPropertyChanged(this.temperatureProperty);
+    this.temperatureProperty.setCachedValueAndNotify(scaleTemperature(temperature,this.config));
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts]: Can you please check if `scaleTemperature` is necessary.
The gateway should do it for you if you specify the correct value for `multipleOf`.

There feature was introduced [here](https://github.com/mozilla-iot/gateway/pull/1971/files#diff-5c4a47e69f3504e8ca1e684755efc3f4R171).

File: src/ruuvitag-adapter.ts [hunk 13]
```diff
@@ -122,17 +125,13 @@ export class RuuviTag extends Device {
...
 
-    this.pressureProperty.setCachedValue(pressure);
-    this.notifyPropertyChanged(this.pressureProperty);
+    this.pressureProperty.setCachedValueAndNotify(scalePressure(pressure,this.config));
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts]: Same here.

File: src/ruuvitag-adapter.ts [hunk 14]
```diff
@@ -122,17 +125,13 @@ export class RuuviTag extends Device {
...
 
-    this.batteryProperty.setCachedValue(batteryVoltage);
-    this.notifyPropertyChanged(this.batteryProperty);
+    this.batteryProperty.setCachedValueAndNotify(batteryVoltage);
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts:134]: And here.

File: manifest.json (hunk 3 above)
Comment (chas-iot) [reply manifest.json:45]: I found that without the minimum and maximum, it allowed users to choose blank/null. 
I'm about to go to bed, so will have to look later in the week.

File: src/ruuvitag-adapter.ts (hunk 1 above)
Comment (chas-iot) [reply src/ruuvitag-adapter.ts]: this is well beyond my knowledge :(

File: manifest.json (hunk 3 above)
Comment (tim-hellhake) [reply manifest.json:45]: Interesting... that doesn't make much sense.

File: src/ruuvitag-adapter.ts (hunk 1 above)
Comment (tim-hellhake) [reply src/ruuvitag-adapter.ts]: No problem at all!
That's why I'm telling you 😄.

File: src/ruuvitag-scaling.ts (hunk 11 above)
Comment (chas-iot) [reply src/ruuvitag-scaling.ts:97]: agreed, but only if the config gets loaded. 
For existing users, on upgrade, I think that the config does not exist. And I needed this fallback while I was struggling with the manifest.id/manifest.name

File: src/ruuvitag-adapter.ts (hunk 12 above)
Comment (chas-iot) [reply src/ruuvitag-adapter.ts]: setCachedValueAndNotify does not alter values, but uses setCachedValue which only normalises booleans and does not touch other values.
And to get the reduction in values written to the database for the Logs screen, it's important to normalise the values before submission. The Logs screen is very slow with a days worth of values with spurious precision

Comment (tim-hellhake) [reply src/ruuvitag-adapter.ts]: Fair point.
I never used the log screen because my logs go directly to influxdb and grafana.

File: src/ruuvitag-adapter.ts [hunk 15]
```diff
@@ -157,6 +155,20 @@ export class RuuviTagAdapter extends Adapter {
...
+    manifest.id = manifest.id || 'ruuvitag-adapter';
+
+    const config = getDefaultConfig();
+    const db = new Database(manifest.id);
+    db.open()
+    .then(() => { return db.loadConfig(); })
+    .then((c) => {
+      mergeLoadedConfig(config,c);
+    })
+    .catch((e) => console.error(e));
```
Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts:170]: There is an even simple way to do this.
```suggestion
    mergeLoadedConfig(config, manifest.config);
```

Comment (tim-hellhake) [inline src/ruuvitag-adapter.ts]: There is an even simpler way to do this.
```suggestion
    mergeLoadedConfig(config, manifest.config);
```

**Comment by chas-iot** [comment]
I think that I have incorporated all of your actionable comments. Please let me know if there is anything that I missed?

File: src/ruuvitag-adapter.ts (hunk 15 above)
Comment (chas-iot) [reply src/ruuvitag-adapter.ts]: manifest.config is undefined - I'll investigate if it's there with another name

Comment (chas-iot) [reply src/ruuvitag-adapter.ts]: It's there as manifest.moziot.config
Which I think you plan to remove

**Comment by chas-iot** [comment]
is there a way to rebuild just the TypeScript without executing the whole package.sh script?
This would save a lot of time in my change/build/test cycle. I'm still learning node and typescript is new to me. And npm is a complete mystery

**Comment by tim-hellhake** [comment]
> is there a way to rebuild just the TypeScript without executing the whole package.sh script?

Sure, just enter `npm run build` on the command-line.
You can execute the tests with `npm test`.
In case you are wondering how npm knows how to build/test the project just look in the `scripts` of the `package.json`.

File: src/ruuvitag-adapter.ts (hunk 15 above)
Comment (tim-hellhake) [reply src/ruuvitag-adapter.ts]: > It's there as manifest.moziot.config

Damn, I forgot the middle part.

> Which I think you plan to remove

What I plan to remove is the `moziot` section in the `package.json`.
Older versions of the gateway (which do not support `setCachedValueAndNotify`) created the `manifest` from the contents of the `package.json`.
Newer gateway versions use the `manifest.json` instead.
In order not to break existing code, the structure remains the same, though.

For reference:
The `manifest.json` is read [here](https://github.com/mozilla-iot/gateway/blob/a526cbf1708856d5e9bd6c84dc6e16a7ff05d0b3/src/addon-manager.js#L614) when the addon is loaded.
The `manifest` is filled [here](https://github.com/mozilla-iot/gateway/blob/a526cbf1708856d5e9bd6c84dc6e16a7ff05d0b3/src/addon-loader.js#L52).

TL;DR; Although `manifest.moziot.config` looks like it comes from the `package.json` it actually comes from the `manifest.json`.

Comment (chas-iot) [reply src/ruuvitag-adapter.ts]: done

**Comment by chas-iot** [comment]
Thanks for the npm commands, simplifies and speeds up my turnaround

**Comment by tim-hellhake** [comment]
Nice, I like it!

image
